### DISCUSSION TITLE
Uploading a local file to the cloud makes TeXstudio think the file was changed

### DISCUSSION THREAD

**Discussion by podkop** [abraunegg/onedrive#1243] [Q&A] [answered]
First of all, thank you very much Contributors for this wonderful app. It works smoothly with multiple folders, and the documentation is clear enough for me as a newbie to Linux.

There is some strange behavior, which is a minor issue for me. It is only observed when using TeXstudio for editing .tex files.
OS is Fedora 33; TeXstudio 3.0.4 is a package installed with dnf.

_Given_: a folder which is configured to automatically sync with OneDrive (as described in advanced-usage.md / part 7). 
1) Open a .tex file located in this folder with TeXstudio.
2) Edit and save the file.
3) TeXstudio immediately opens a dialog window with a warning that the file was changed from outside of TeXstudio.

This does not happen when I edit .tex files in Dropbox synced folders.

Using `stat`, I checked what happens with a .tex file after I change it, and then enable synchronization. 
* For a OneDrive synced folder, after I do `systemctl --user start onedrive-my-config`, only _access time_ is changed.
* For a Dropbox synced folder, after I enable Dropbox synchronization, both _access time_ and _change time_ change.

I do not know if it is relevant, but files in OneDrive folder have permissions 600, while in the Dropbox folder the permissions are 644.

I wonder what does OneDrive do to a file, which makes it look like it was changed from outside. There is also a topic in Stackexchange related to this situation https://tex.stackexchange.com/questions/575945/how-to-enable-silent-reloading-in-texstudio

**Comment by abraunegg** [comment, marked as answer]
@podkop 
Unless you are building your client from source, the Fedora 33 package is old and unsupported (onedrive-2.4.8-1.fc33) - what client version are you using?

> TeXstudio immediately opens a dialog window with a warning that the file was changed from outside of TeXstudio

Once a file is uploaded to OneDrive, depending on your account type, files can be uploaded, then OneDrive *modifies* the file by adding metadata to this file, fundamentally changing this file. To keep the file now in-sync, the local file is updated with this now *modified* file.

The only way to see *exactly* what is going on for you is to generate a verbose debug log as per this [process](https://github.com/abraunegg/onedrive/wiki/Generate-debug-log-for-support). 

It could also be that TeXstudio is 'broken' - as it has been found in the past with other applications on Linux, the process by which they write files (new or modified) is not ideal.

> I do not know if it is relevant, but files in OneDrive folder have permissions 600, while in the Dropbox folder the permissions are 644.

You can change the default permissions as per: https://github.com/abraunegg/onedrive/blob/master/docs/USAGE.md#sync_dir-directory-and-file-permissions

The default of 600 is there to ensure maximum user security on the files - they are your files, not group or world readable. Changing to 644 similar to Dropbox I do not think will have an impact, unless TeXstudio is expecting 644 and now cannot read the file correctly due to no group / world user access. You can test this with a file that is not being synced with OneDrive by just setting `chmod 0600` on that file and see how TeXstudio reacts.

**Reply by podkop** [reply]
Thank you very much for the reply. Indeed, I use 2.4.8-1.fc33. I prefer to install everything via dnf in order to not "break" anything. I am not an experienced user, so if some app is included in dnf, for me it sounds like a most reliable version.

The standard Gnome text editor (gedit 3.38.1) has the same behavior: when a folder is synchronized with OneDrive, after I save a text file and try to close the editor, it shows a warning that the file was changed from outside. 

It is strange that despite OneDrive changes file metadata, "change time" remains the same. It is also strange that despite for Dropbox-synced files, "change time" changes, the system does not think that the file was changed.

Below are logs related to uploading changes of a text file ADM3.out

\# 627 - \# 633
```
Processing General/IEEE_Access/Source/ADM3.out
[DEBUG] skip_file evaluation for: ADM3.out
The file last modified time has changed
The file content has changed
Uploading modified file General/IEEE_Access/Source/ADM3.out ... skipped.
Skip Reason: Microsoft Sharepoint 'enrichment' after upload issue
See: https://github.com/OneDrive/onedrive-api-docs/issues/935 for further details
```

\# 937 - \# 948
```
[DEBUG] Checking file: ./General/IEEE_Access/Source/ADM3.out
[DEBUG] skip_file evaluation for: /General/IEEE_Access/Source/ADM3.out
[DEBUG] File upload destination is users default driveId ..
[DEBUG] Request URL = https://graph.microsoft.com/v1.0/drives/<drive_ID>/root:/.%2FGeneral%2FIEEE_Access%2FSource%2FADM3.out?select=id,name,eTag,cTag,deleted,file,folder,root,fileSystemInfo,remoteItem,parentReference,size
Requested file to upload exists on OneDrive - local database is out of sync for this file: ./General/IEEE_Access/Source/ADM3.out
Requested file to upload is newer than existing file on OneDrive
Uploading modified file ./General/IEEE_Access/Source/ADM3.out ...  done.
[DEBUG] lastModifiedDateTime (OneDrive item): 2021-Jan-30 11:53:46Z
[DEBUG] Adding to database: Item("<drive_ID>", "<some_code>", "ADM3.out", file, "\"{<another_code>},15\"", "\"c:{<another_code>},12\"", 2021-Jan-30 11:53:46Z, "<some_code>", "", "", "<some_code>=", "", "", "Y")
Due to Microsoft Sharepoint 'enrichment' of files, downloading 'enriched' file to ensure local file is in-sync
See: https://github.com/OneDrive/onedrive-api-docs/issues/935 for further details
[DEBUG] Setting file permissions for: ./General/IEEE_Access/Source/ADM3.out
```
\# 994 - \# 1009
```
[DEBUG] inotify event flag: IN_ALL_EVENTS
[DEBUG] inotify path event for: General/IEEE_Access/Source/ADM3.out
[DEBUG] skip_file evaluation for: General/IEEE_Access/Source/ADM3.out
[DEBUG] event IN_CLOSE_WRITE and ...: General/IEEE_Access/Source/ADM3.out
[M] File changed: General/IEEE_Access/Source/ADM3.out
Uploading differences of General/IEEE_Access/Source/ADM3.out
[DEBUG] Processing DB entries for this driveId: <drive_ID>
Processing General/IEEE_Access/Source/ADM3.out
[DEBUG] skip_file evaluation for: ADM3.out
The file has not changed
Uploading new items of General/IEEE_Access/Source/ADM3.out
[DEBUG] Checking file: General/IEEE_Access/Source/ADM3.out
[DEBUG] skip_file evaluation for: General/IEEE_Access/Source/ADM3.out
[DEBUG] Final True-Up: Perform a full walk of OneDrive objects because a full scan was required
[DEBUG] Calling sync.applyDifferences(true);
Applying changes of Path ID: <path_ID>
```

**Comment by abraunegg** [comment]
@podkop 
> It is strange that despite OneDrive changes file metadata, "change time" remains the same.

However the file *is* different - if you compare the files, you will see subtle differences. This is why your application is complaining that it was modified outside of the application - because it was modified by SharePoint.

Not an issue or bug with the client, but with SharePoint as per the GitHub issue linked in the debug.

**Reply by podkop** [reply]
The mentioned issue does not concern text files, and does not concern business accounts. The behavior I observe concerns text files, and I use a business account. I applied `diffoscope` to compare a file before and after sync - no difference.

The only thing OneDrive client does to a file is sets file permissions.

**Reply by abraunegg** [reply]
@podkop 
> The mentioned issue does not concern text files, and does not concern business accounts. The behavior I observe concerns text files, and I use a business account. 

You appear to be saving files on a Microsoft SharePoint site. You can only do this if you are using a business account - so - this issue *is* impacting you, because of the way Microsoft mangles data when saving *any* file on SharePoint.

> The only thing OneDrive client does to a file is sets file permissions.

As per above, you can change the default permissions as per: https://github.com/abraunegg/onedrive/blob/master/docs/USAGE.md#sync_dir-directory-and-file-permissions

This will determine if this is the root cause of your issue.

**Reply by abraunegg** [reply]
@podkop 
Please can you also review PR #1250

What this will do is compare the file hash pre/post upload before determine if the file needs to be downloaded post SharePoint modification.

```
git clone https://github.com/abraunegg/onedrive.git
cd onedrive
git fetch origin pull/1250/head:pr1250
git checkout pr1250
./configure; make clean; make;
```
To run the PR, you need to run the client from the PR build directory:
```
./onedrive <any options needed>
```

When running the PR, your version should be: `onedrive v2.4.9-12-g43bcf60` or greater.

**Reply by podkop** [reply]
Thank you very much for update. I'm neither a programmer nor an admin, so all those manipulations are new for me. I will try to do it, but it may take some time.
